pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,159 @@
+    >>>     cost_rotosel.append(cost(x, generators))
+    >>>     x, generators = opt.step(cost, x, generators)
+
+    The optimized values for x should now be stored in ``x`` together with the optimal gates for
+    the circuit, while steps-vs-cost can be seen by plotting ``cost_rotosel``.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, **kwargs):
+        self.possible_generators = kwargs.get("possible_generators", [qml.RX, qml.RY, qml.RZ])
+
+    def step(self, objective_fn, x, generators):
+        """Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): The objective function for optimization. It should take two lists
+                of the values ``x`` and the gates ``generators`` as inputs, and return a single value.
+            x (array[float]): NumPy array containing the initial values of the variables to be optimized over.
+            generators (list[~.Operation]): List containing the initial ``pennylane.ops.qubit``
+                operators to be used in the circuit and optimized over.
+
+        Returns:
+            array: The new variable values :math:`x^{(t+1)}` as well as the new generators.
+        """
+        # make sure that x is an array
+        if np.ndim(x) == 0:
+            x = np.array([x])
+
+        for d, _ in enumerate(x):
+            x[d], generators[d] = self._find_optimal_params(objective_fn, x, generators, d)
+
+        return x, generators
+
+    def _find_optimal_params(self, objective_fn, x, generators, d):
+        """Optimizer for the generators.
+
+        Optimizes for the best generator at position ``d``.
+
+        Args:
+            objective_fn (function): The objective function for optimization. It should take two lists
+                of the values ``x`` and the gates ``generators`` as inputs, and return a single value.
+            x (array[float]): NumPy array containing the initial values of the variables to be optimized over.
+            generators (list[~.Operation]): List containing the initial ``pennylane.ops.qubit``
+                operators to be used in the circuit and optimized over.
+            d (int): The position in the input array ``x`` containing the value to be optimized.
+
+        Returns:
+            tuple: Tuple containing the parameter value and generator that, at position ``d`` in their
+            respective lists ``x`` and ``generators``, optimizes the objective function.
+        """
+        for i, generator in enumerate(self.possible_generators):
+            generators[d] = generator
+
+            x = self._rotosolve(objective_fn, x, generators, d)
+            params_cost = objective_fn(x, generators)
+
+            # initialize optimal generator with first item in list and update if necessary
+            if i == 0 or params_cost <= params_opt_cost:
+                params_opt_d = x[d]
+                params_opt_cost = params_cost
+                generators_opt_d = generator
```

Comment:
Minor suggestion (not too crucial): here the logic of initializing the parameters and then updating once a more optimal cost was found is slightly blurred. One way to update this could be for example by using a collection, which at first contains every generator and using `pop` to get one generator at a time.

Reply:
Thanks for clarifying this @antalszava. I think it looks better now, when I define the initial values before the loop, and then only update ``params_opt_cost``, ``params_opt_cost`` and ``generators_opt_d`` when better values are found. Please tell me what you think!